torch/nn/functional.py

```diff
@@ -5175,19 +5175,20 @@ def multi_head_attention_forward(
     # (deep breath) calculate attention and out projection
     #
 
-    B, Nt, E = q.shape
-    q_scaled = q / math.sqrt(E)
     if attn_mask is not None:
-        attn_output_weights = torch.baddbmm(attn_mask, q_scaled, k.transpose(-2, -1))
-    else:
-        attn_output_weights = torch.bmm(q_scaled, k.transpose(-2, -1))
-    attn_output_weights = softmax(attn_output_weights, dim=-1)
-    if dropout_p > 0.0:
-        attn_output_weights = dropout(attn_output_weights, p=dropout_p)
+        if attn_mask.size(0) == 1:
```

Comment:
hacky.. the API between SDP vs the nn.funcitonal.mha has a decent amount of impedance mismatch requiring all this transposing and viewing fluff. Ideally this would also work with nested tensors out of the box but more thought needs to be takin to do so